compiler/rustc_parse/src/parser/item.rs

```diff
@@ -670,14 +670,44 @@ impl<'a> Parser<'a> {
             }
             match parse_item(self) {
                 Ok(None) => {
+                    let is_unnecessary_semicolon = !items.is_empty()
+                        // When the close delim is `)` in a case like the following, `token.kind` is expected to be `token::CloseDelim(Delimiter::Parenthesis)`,
+                        // but the actual `token.kind` is `token::CloseDelim(Delimiter::Bracket)`.
+                        // This is because the `token.kind` of the close delim is treated as the same as
+                        // that of the open delim in `TokenTreesReader::parse_token_tree`, even if the delimiters of them are different.
+                        // Therefore, `token.kind` should not be compared here.
+                        //
+                        // issue-60075.rs
+                        // ```
+                        // trait T {
+                        //     fn qux() -> Option<usize> {
+                        //         let _ = if true {
+                        //         });
+                        //          ^ this close delim
+                        //         Some(4)
+                        //     }
+                        // ```
+                        && self
+                            .span_to_snippet(self.prev_token.span)
+                            .map_or(false, |snippet| snippet == "}")
```

Comment:
Why do we need `span_to_snippet` here? Can you just check `prev_token.kind`?

Reply:
The comments above explain why.

When the close delim is `)` in a case like the following, `token.kind` is expected to be `token::CloseDelim(Delimiter::Parenthesis)`, but the actual `token.kind` is `token::CloseDelim(Delimiter::Bracket)`. This is because the `token.kind` of the close delim is treated as the same as that of the open delim in `TokenTreesReader::parse_token_tree`, even if the delimiters of them are different. Therefore, `token.kind` should not be compared here.

issue-60075.rs
```
trait T {
     fn qux() -> Option<usize> {
         let _ = if true {
         });
          ^ this close delim
         Some(4)
     }
```

Reply:
I totally did not read that comment. Thank you for clarifying.